Deduplicate file downloads (fixes #234)

jessebrennan opened the pull request:
Adds another download command `download_manifest_v2` which downloads with the new version format.

jessebrennan commented:
@hannes-ucsc I think this is a flaky test:

```
======================================================================
ERROR: test_post_search_cli (test.integration.dss.test_dss_cli.TestDssCLI)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/HumanCellAtlas/dcp-cli/test/integration/dss/test_dss_cli.py", line 26, in test_post_search_cli
    hca.cli.main(args)
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/cli.py", line 117, in main
    result = parsed_args.entry_point(parsed_args)
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 163, in _cli_call
    return self._consume_response(self._request(vars(cli_args)))
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 148, in _request
    raise SwaggerAPIException(response=res)
hca.util.exceptions.SwaggerAPIException: internal_server_error: Elasticsearch: 10 of 130 shards failed. (HTTP 500). Details:
Traceback (most recent call last):
  File "/var/task/chalicelib/dss/error.py", line 59, in wrapper
    return func(*args, **kwargs)
  File "/var/task/chalicelib/dss/api/search.py", line 75, in post
    f"Elasticsearch: {page['_shards']['failed']} "
dss.error.DSSException
----------------------------------------------------------------------
```

Looking into how to fix it. Otherwise tests are passing.

DailyDreaming commented:
@jessebrennan It is not.  Pull from master and it'll be fixed.

DailyDreaming commented on hca/dss/__init__.py:
```diff
@@ -100,89 +102,184 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+    def _download_file(self,
+                       file_uuid,
+                       file_sha256,
+                       dest_path,
```
Nit: `dest_path` shouldn't be renamed to `file_path` as it's a file and not a destination.

DailyDreaming commented on hca/dss/__init__.py:
```diff
@@ -33,10 +34,11 @@ class DSSClient(SwaggerClient):
     Client for the Data Storage Service API.
     """
     UPLOAD_BACKOFF_FACTOR = 1.618
+    FILE_NESTING_PARAMETER = [2, 4]
```
`FILE_NESTING_PARAMETER` tells the CLI where to put slashes to make a directory structure using the checksum.  There should be a comment about that, the naming is a little cryptic.

DailyDreaming commented on hca/dss/__init__.py:
```diff
@@ -107,87 +109,182 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+    @classmethod
+    def _file_path(cls, checksum):
+        """
+        returns a file's relative local path based on the nesting parameters and the files hash
```
Comment could be more descriptive.  As in, given a string: `64f0b864f0b8e7fc22a8c5f7ec6649530a03e4` it will return `64/f0b8/64f0b8e7fc22a8c5f7ec6649530a03e4` to be used as a directory structure for the file, with the slashes always where `FILE_NESTING_PARAMETER` indicates.

Also, will the user be getting a ton of obscurely named (from there point of view) directories?  If so, should we rethink this?

Also, why create multiple directories?  Why not just one directory with the checksum name, as in `FILE_NESTING_PARAMETER=[len(checksum)]`?

DailyDreaming replied:
Is the epic #232?  I'm having trouble seeing much information there.  I'd still like to argue for simplifying this by not creating the extra dirs if that's not too unreasonable.  I really like simple code when one can get away with it.

hannes-ucsc replied:
Yeah, the epic doesn't make the plan clear, sorry. I talked about this at the F2F, though, and it was discussed at length on Slack in `#cli-download-layout`. Feel free to talk to me personally, if you have more questions about the big picture. 

The point is that we want to maintain a content-addressed file store that could hold hundreds of thousands of files. We can't put them all in the same directory because that would mean even running `ls` will take minutes to complete.

DailyDreaming commented on hca/dss/__init__.py:
```diff
@@ -107,87 +109,182 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+        """
+
+        Process the given manifest file in TSV (tab-separated values) format and download the files referenced by it.
+        The files are downloaded in the version 2 format.
```
What is the "version 2 format" and can you include a link to it?

DailyDreaming replied:
Ah, that sounds reasonable.  Just confusing to me seeing this new.

jessebrennan replied:
@hannes-ucsc I need some clarification on this plan before I write a description.

DailyDreaming replied:
Will wait for this then.  Otherwise looks good to approve.

DailyDreaming commented on hca/dss/__init__.py:
```diff
@@ -107,87 +109,182 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+        :param checksum: a string checksum
+        :return: relative Path object
+        """
+        assert(sum(cls.FILE_NESTING_PARAMETER) <= len(checksum))
```
This assert will fail if the FILE_NESTING_PARAMETER is [15, 18] and the checksum length is 32, even though that seems perfectly valid to me.

DailyDreaming replied:
Would you need to worry about things like `[2, 2, 2]`?  I don't imagine this needs to be configured ever, and I'm kind of arguing for not needing it at all above, but I was imagining something more like:
```
i = sorted(cls.FILE_NESTING_PARAMETER)
assert((i[-1] <= len(checksum)) and i[0] >= 0 and len(set(i)) == len(i))
```
Please correct me if I'm misunderstanding.

hannes-ucsc replied:
`[2, 2, 2]` is fine. And yes, I think you are misunderstanding. AFAICT, @jessebrennan's implementation is correct and straight-forward.

DailyDreaming commented on hca/dss/__init__.py:
```diff
@@ -107,87 +109,182 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+            # This first check avoids downloading the file unnecessarily
+            if os.path.exists(path):
+                logger.info("File %s already present. Skipping download.", file_uuid)
+                return
```
If `for row in rows:` is downloading a different file for each row, won't this `return` skip the rest of the downloads as soon as you hit a duplicated file?

hannes-ucsc replied:
The tests should have caught this. Have the tests remove two files after the first invocation of the code under test then invoke again. Then check for completeness. With this bug in place only one file would have been replaced.

In general, prefer `else` over bail-outs like `continue` or `return`. Sometimes they can be useful, but usually only in long functions that are already deeply nested. But those have a bad smell may need to be refactored.

jessebrennan replied:
Thanks for catching this Lon. That would have been baaaad.

diekhans commented:
general comment: I think you will find code much easier to write and understand if you write more, smaller functions.  I usually go from 10-20 line functions.  If at all possible, the bodies of loops are usually function calls if more than 2 -3 lines.   The complex flow of control with continues and breaks can be made to mostly go away.   With small, internal functions, code becomes much clearly, because the function name being called is also documentation.

anyway, that is the most useful thing I have learned in decades of programming.

DailyDreaming commented:
@diekhans I wholeheartedly agree with that style of coding.  Though `git` is doing some weird history stuff here, so a lot of the changes it's seeing in this PR were already there previously.

diekhans commented:
@DailyDreaming better to learn from someone else's mistakes than
your own ;-)  I find that refactoring large functions often makes
it much easier to add new code and to understand if you did the
right thing.

One certain genome browser has a function which is like 600
lines long with deep loop/if nesting, breaks, and continues.

I just closed the file, I couldn't look.

Lon Blauvelt <notifications@github.com> writes:
> @diekhans I wholeheartedly agree with that style of coding. Though git is doing
> some weird history stuff here, so a lot of the changes it's seeing in this PR
> were already there previously.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

DailyDreaming commented:
@diekhans lol, say no more.

jessebrennan commented:
@diekhans Trust me, I would not have written that function myself. I considered refactoring it, but then I would have wanted to make unit tests for it and it's quite complicated and i didn't want to break it and it seemed to work well as is, so I left it. :man_shrugging:

jessebrennan commented on hca/dss/__init__.py:
```diff
@@ -107,87 +111,181 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+        return os.path.join(*path_pieces)
+
+    @classmethod
+    def _parse_manifest(cls, manifest):
```
The way I wrote this, the entire file is read into memory at once. Will we ever encounter a manifest large enough for this to be a problem?

hannes-ucsc replied:
I doubt it.

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,75 @@
+
+
+def _touch_file(*args, **kwargs):
+    path = args[2] if len(args) > 2 else args[0]
```
Explain!

jessebrennan replied:
Done

hannes-ucsc replied:
I see now. Creating a wrapper function for the mock case would be cleared and eliminate the need for an explanation.

jessebrennan replied:
done.

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,75 @@
+    try:
+        os.makedirs(os.path.split(path)[0])
+    except OSError:
+        # dirs already exist
```
Not always true, I think:

```
>>> import os
>>> os.makedirs("/x")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/x'
>>> os.makedirs("/")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode)
IsADirectoryError: [Errno 21] Is a directory: '/'
>>> os.makedirs("/tmp/foo")
>>> os.makedirs("/tmp/foo")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: '/tmp/foo'
```

jessebrennan replied:
Good point. I know on 2.7 FileExistsError doesn't exist, but I can still check the error string in a portable way.

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,75 @@
+        self.dss.download_manifest_v2('manifest.tsv', 'aws')
+        self.assertEqual(download_func.call_count, len(self.manifest) - 1)
+        # The check above is good, but we also need to check how many files were actually writen
+        num_files = sum(map(lambda tup: len(tup[2]), os.walk('.')))
```
I'd like this to be stronger and assert everything, file names and paths. Not just counts. Also 

```suggestion
        num_files = sum(len(files) for path, dirs, files in os.walk('.'))
```

map() is rarely more readable than a comprehension. And shorter, too. Mine is only longer because I mention the other two tuple elements. Even the version that doesn't name them is more readable, and on par re length.

```suggestion
        num_files = sum(len(files) for _, _, files in os.walk('.'))
```

jessebrennan replied:
Yes, this is clearly superior.

DailyDreaming replied:
I'd use scandir instead of `os.walk()` if you haven't heard of it yet.  It's a dependency already in `dcp-cli` and it's a part of the os standard library in python3.  It's about 2-20 times faster than walk and seems strictly better when used: https://www.python.org/dev/peps/pep-0471/  The change to the code would be:
`num_files = sum(len(files) for _, _, files in scandir.walk('.'))`

jessebrennan replied:
Good to know, thanks.

DailyDreaming replied:
It's an order of magnitude increase in speed, which is what the epic is all about.  If you want a blunt experience of it, try it from your home dir vs. `os.walk()`.

jessebrennan replied:
I believe you, that's why I changed it.

hannes-ucsc commented:
@jessebrennan, Are you concerned about the lack in patch coverage?

jessebrennan commented:
@hannes-ucsc no. The low patch coverage is because I refactored the mechanism used for writing files from bundle download implementation. All uncovered patch changes are from this code. Since I changed the indent of these lines coverage credits them to my patch.

That function *should* be covered in tests, but I argue in a separate issue.

hannes-ucsc commented:
> That function should be covered in tests, but I argue in a separate issue.

Sounds good. Could you create that ticket, please?

hannes-ucsc commented on hca/dss/__init__.py:
```diff
@@ -107,87 +111,185 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+            path = self._file_path(sha)
+            version = row['file_version']
+            size = row['file_size']
+            # This first check avoids downloading the file unnecessarily
```
I think this is obvious, especially considering the log message in the `then` branch.

jessebrennan replied:
good point.

diekhans commented:
I think I accidentally resolved Hannes' comment on the makedir/OSExrror.  There doesn't appear to be an undo :-(

My point was that OSError could also be called on things like permission denied, so assuming it is directory exist is wrong.  While this will cause failures later on,  it isn't the best approach to make this assumption.  You can check the OSError.errno field to determine if it is the error expected.  

But it is clear and less lines of code to just do

`
  os.makedirs(name, exist_ok=True)
`

hannes-ucsc commented:
@diekhans,  `exist_ok` isn't available on 2.7.

diekhans commented:
I didn't realize you were stuck on 2.7, my sympathy.
IMHO, Not worth trying to make easier, it will fail somewhere else
if it is permission denied, at real file, etc

Hannes Schmidt <notifications@github.com> writes:
> @diekhans, exist_ok isn't available on 2.7.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

jessebrennan commented:
Weird flaky test failures only happened once on Python2.7 build...

```
======================================================================
ERROR: test_get_files_cli (test.integration.dss.test_dss_cli.TestDssCLI)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/HumanCellAtlas/dcp-cli/test/integration/dss/test_dss_cli.py", line 44, in test_get_files_cli
    hca.cli.main(args=upload_args)
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/cli.py", line 131, in main
    result = parsed_args.entry_point(parsed_args)
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 544, in arg_forwarder
    return command(**command_args)
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/dss/__init__.py", line 433, in upload
    files=file_args)
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 202, in __call__
    return self._consume_response(self._request(kwargs))
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 190, in _request
    raise SwaggerAPIException(response=res)
SwaggerAPIException: Unauthorized: Authorization token is invalid (HTTP 401). Details:
Traceback (most recent call last):
  File "/var/task/chalicelib/dss/util/security.py", line 69, in verify_jwt
    algorithms=allowed_algorithms,
  File "/var/task/jwt/api_jwt.py", line 104, in decode
    self._validate_claims(payload, merged_options, **kwargs)
  File "/var/task/jwt/api_jwt.py", line 140, in _validate_claims
    self._validate_aud(payload, audience)
  File "/var/task/jwt/api_jwt.py", line 204, in _validate_aud
    raise InvalidAudienceError('Invalid audience')
jwt.exceptions.InvalidAudienceError: Invalid audience
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/var/task/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/task/connexion/decorators/decorator.py", line 73, in wrapper
    response = function(request)
  File "/var/task/connexion/decorators/security.py", line 113, in wrapper
    token_info = token_info_func(token)
  File "/var/task/chalicelib/dss/util/security.py", line 74, in verify_jwt
    raise DSSException(401, 'Unauthorized', 'Authorization token is invalid') from ex
dss.error.DSSException
======================================================================
ERROR: test_python_bindings (test.integration.dss.test_dss_api.TestDssApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/HumanCellAtlas/dcp-cli/test/integration/dss/test_dss_api.py", line 213, in test_python_bindings
    bundle_output = client.upload(src_dir=bundle_path, replica="aws", staging_bucket=self.staging_bucket)
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/dss/__init__.py", line 393, in upload
    source_url=source_url
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 190, in _request
    raise SwaggerAPIException(response=res)
SwaggerAPIException: Unauthorized: Authorization token is invalid (HTTP 401). Details:
Traceback (most recent call last):
  File "/var/task/chalicelib/dss/util/security.py", line 69, in verify_jwt
    algorithms=allowed_algorithms,
  File "/var/task/jwt/api_jwt.py", line 104, in decode
    self._validate_claims(payload, merged_options, **kwargs)
  File "/var/task/jwt/api_jwt.py", line 140, in _validate_claims
    self._validate_aud(payload, audience)
  File "/var/task/jwt/api_jwt.py", line 204, in _validate_aud
    raise InvalidAudienceError('Invalid audience')
jwt.exceptions.InvalidAudienceError: Invalid audience
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/var/task/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/task/connexion/decorators/decorator.py", line 73, in wrapper
    response = function(request)
  File "/var/task/connexion/decorators/security.py", line 113, in wrapper
    token_info = token_info_func(token)
  File "/var/task/chalicelib/dss/util/security.py", line 74, in verify_jwt
    raise DSSException(401, 'Unauthorized', 'Authorization token is invalid') from ex
dss.error.DSSException
======================================================================
ERROR: test_python_login_logout_service_account (test.integration.dss.test_dss_api.TestDssApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/HumanCellAtlas/dcp-cli/test/__init__.py", line 66, in save_and_restore_tweak_config
    f(*args, **kwargs)
  File "/home/travis/build/HumanCellAtlas/dcp-cli/test/integration/dss/test_dss_api.py", line 322, in test_python_login_logout_service_account
    resp = client.put_subscription(es_query=query, callback_url="https://www.example.com", replica="aws")
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 202, in __call__
    return self._consume_response(self._request(kwargs))
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 190, in _request
    raise SwaggerAPIException(response=res)
SwaggerAPIException: Unauthorized: Authorization token is invalid (HTTP 401). Details:
Traceback (most recent call last):
  File "/var/task/chalicelib/dss/util/security.py", line 69, in verify_jwt
    algorithms=allowed_algorithms,
  File "/var/task/jwt/api_jwt.py", line 104, in decode
    self._validate_claims(payload, merged_options, **kwargs)
  File "/var/task/jwt/api_jwt.py", line 140, in _validate_claims
    self._validate_aud(payload, audience)
  File "/var/task/jwt/api_jwt.py", line 204, in _validate_aud
    raise InvalidAudienceError('Invalid audience')
jwt.exceptions.InvalidAudienceError: Invalid audience
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/var/task/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/task/connexion/decorators/decorator.py", line 73, in wrapper
    response = function(request)
  File "/var/task/connexion/decorators/security.py", line 113, in wrapper
    token_info = token_info_func(token)
  File "/var/task/chalicelib/dss/util/security.py", line 74, in verify_jwt
    raise DSSException(401, 'Unauthorized', 'Authorization token is invalid') from ex
dss.error.DSSException
======================================================================
ERROR: test_python_manifest_download (test.integration.dss.test_dss_api.TestDssApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/HumanCellAtlas/dcp-cli/test/integration/dss/test_dss_api.py", line 144, in test_python_manifest_download
    staging_bucket=self.staging_bucket)
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/dss/__init__.py", line 393, in upload
    source_url=source_url
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 190, in _request
    raise SwaggerAPIException(response=res)
SwaggerAPIException: Unauthorized: Authorization token is invalid (HTTP 401). Details:
Traceback (most recent call last):
  File "/var/task/chalicelib/dss/util/security.py", line 69, in verify_jwt
    algorithms=allowed_algorithms,
  File "/var/task/jwt/api_jwt.py", line 104, in decode
    self._validate_claims(payload, merged_options, **kwargs)
  File "/var/task/jwt/api_jwt.py", line 140, in _validate_claims
    self._validate_aud(payload, audience)
  File "/var/task/jwt/api_jwt.py", line 204, in _validate_aud
    raise InvalidAudienceError('Invalid audience')
jwt.exceptions.InvalidAudienceError: Invalid audience
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/var/task/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/task/connexion/decorators/decorator.py", line 73, in wrapper
    response = function(request)
  File "/var/task/connexion/decorators/security.py", line 113, in wrapper
    token_info = token_info_func(token)
  File "/var/task/chalicelib/dss/util/security.py", line 74, in verify_jwt
    raise DSSException(401, 'Unauthorized', 'Authorization token is invalid') from ex
dss.error.DSSException
======================================================================
ERROR: test_python_nested_bundle_upload_download (test.integration.dss.test_dss_api.TestDssApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/HumanCellAtlas/dcp-cli/test/integration/dss/test_dss_api.py", line 68, in test_python_nested_bundle_upload_download
    staging_bucket=self.staging_bucket)
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/dss/__init__.py", line 393, in upload
    source_url=source_url
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 190, in _request
    raise SwaggerAPIException(response=res)
SwaggerAPIException: Unauthorized: Authorization token is invalid (HTTP 401). Details:
Traceback (most recent call last):
  File "/var/task/chalicelib/dss/util/security.py", line 69, in verify_jwt
    algorithms=allowed_algorithms,
  File "/var/task/jwt/api_jwt.py", line 104, in decode
    self._validate_claims(payload, merged_options, **kwargs)
  File "/var/task/jwt/api_jwt.py", line 140, in _validate_claims
    self._validate_aud(payload, audience)
  File "/var/task/jwt/api_jwt.py", line 204, in _validate_aud
    raise InvalidAudienceError('Invalid audience')
jwt.exceptions.InvalidAudienceError: Invalid audience
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/var/task/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/task/connexion/decorators/decorator.py", line 73, in wrapper
    response = function(request)
  File "/var/task/connexion/decorators/security.py", line 113, in wrapper
    token_info = token_info_func(token)
  File "/var/task/chalicelib/dss/util/security.py", line 74, in verify_jwt
    raise DSSException(401, 'Unauthorized', 'Authorization token is invalid') from ex
dss.error.DSSException
======================================================================
ERROR: test_python_subscriptions (test.integration.dss.test_dss_api.TestDssApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/HumanCellAtlas/dcp-cli/test/integration/dss/test_dss_api.py", line 255, in test_python_subscriptions
    resp = client.put_subscription(es_query=query, callback_url="https://www.example.com", replica="aws")
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 202, in __call__
    return self._consume_response(self._request(kwargs))
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 190, in _request
    raise SwaggerAPIException(response=res)
SwaggerAPIException: Unauthorized: Authorization token is invalid (HTTP 401). Details:
Traceback (most recent call last):
  File "/var/task/chalicelib/dss/util/security.py", line 69, in verify_jwt
    algorithms=allowed_algorithms,
  File "/var/task/jwt/api_jwt.py", line 104, in decode
    self._validate_claims(payload, merged_options, **kwargs)
  File "/var/task/jwt/api_jwt.py", line 140, in _validate_claims
    self._validate_aud(payload, audience)
  File "/var/task/jwt/api_jwt.py", line 204, in _validate_aud
    raise InvalidAudienceError('Invalid audience')
jwt.exceptions.InvalidAudienceError: Invalid audience
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/var/task/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/task/connexion/decorators/decorator.py", line 73, in wrapper
    response = function(request)
  File "/var/task/connexion/decorators/security.py", line 113, in wrapper
    token_info = token_info_func(token)
  File "/var/task/chalicelib/dss/util/security.py", line 74, in verify_jwt
    raise DSSException(401, 'Unauthorized', 'Authorization token is invalid') from ex
dss.error.DSSException
======================================================================
ERROR: test_python_upload_download (test.integration.dss.test_dss_api.TestDssApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/HumanCellAtlas/dcp-cli/test/integration/dss/test_dss_api.py", line 87, in test_python_upload_download
    staging_bucket=self.staging_bucket)
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/dss/__init__.py", line 393, in upload
    source_url=source_url
  File "/home/travis/build/HumanCellAtlas/dcp-cli/hca/util/__init__.py", line 190, in _request
    raise SwaggerAPIException(response=res)
SwaggerAPIException: Unauthorized: Authorization token is invalid (HTTP 401). Details:
Traceback (most recent call last):
  File "/var/task/chalicelib/dss/util/security.py", line 69, in verify_jwt
    algorithms=allowed_algorithms,
  File "/var/task/jwt/api_jwt.py", line 104, in decode
    self._validate_claims(payload, merged_options, **kwargs)
  File "/var/task/jwt/api_jwt.py", line 140, in _validate_claims
    self._validate_aud(payload, audience)
  File "/var/task/jwt/api_jwt.py", line 204, in _validate_aud
    raise InvalidAudienceError('Invalid audience')
jwt.exceptions.InvalidAudienceError: Invalid audience
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/var/task/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/task/connexion/decorators/decorator.py", line 73, in wrapper
    response = function(request)
  File "/var/task/connexion/decorators/security.py", line 113, in wrapper
    token_info = token_info_func(token)
  File "/var/task/chalicelib/dss/util/security.py", line 74, in verify_jwt
    raise DSSException(401, 'Unauthorized', 'Authorization token is invalid') from ex
dss.error.DSSException
```

hannes-ucsc commented:
Re flaky test. Might have had to do with DSS deployment.

https://humancellatlas.slack.com/archives/G9XD6L0AD/p1552411754145300

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,78 @@
+        self.dss.download_manifest_v2('manifest.tsv', 'aws')
+        self.assertEqual(download_func.call_count, len(self.manifest) - 1)
+        # The check above is good, but we also need to check how many files were actually writen
+        num_files = sum(len(files) for _, _, files in scandir.walk('.'))
```
I originally mentioned this in https://github.com/HumanCellAtlas/dcp-cli/pull/269#discussion_r264769096 (first sentence). Consider asserting the actual set of files rather than just its cardinality. Let me know if that would be too difficult, and why.

jessebrennan replied:
Woops! I overlooked that. It's now done.

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,82 @@
+
+    def _assert_all_files_downloaded(self):
+        num_files = sum(len(files) for _, _, files in scandir.walk('.'))
+        self.assertEqual(num_files, len(self.manifest))  # -1 for header, but +1 for manifest.tsv in cwd
```
Assert that the set of files on disk contain that specific files. That is stronger and eliminates the need to explain.

jessebrennan replied:
Nothing is stronger (but maybe as strong) as asserting that every file in the manifest is downloaded and that there is the same number of files. I left the count to make sure that the sets are the same size and thus equivalent.

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,82 @@
+    def _assert_all_files_downloaded(self):
+        num_files = sum(len(files) for _, _, files in scandir.walk('.'))
+        self.assertEqual(num_files, len(self.manifest))  # -1 for header, but +1 for manifest.tsv in cwd
+        expected_file_paths = [self.dss._file_path(row[4]) for row in self.manifest[1:]]
```
We shouldn't rely on code under test to compute an assertion. If there is a bug in the code under test, the assertion will be wrong. I want to see explicit paths like `3f/4c6d/3f4c6d…` here in a literal assertion. I'd rather have a test with tighter coupling to the code under test, and that is therefore more brittle (false negatives if the code under test is changed) than a "smart" assertion that risks giving false positives if the code under test is changed.

jessebrennan replied:
This is a better approach, but the '/'s aren't portable (right?) I could write a bunch of path.joins to get the file's path, but then I end up basically writing `_file_path()` all over again. It seems like it's not worth it. What do you think?

jessebrennan replied:
Nvm, I'll just use `os.sep` and a lot of plus signs.

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,82 @@
+        self.assertEqual(num_files, len(self.manifest))  # -1 for header, but +1 for manifest.tsv in cwd
+        expected_file_paths = [self.dss._file_path(row[4]) for row in self.manifest[1:]]
+        for path in expected_file_paths:
+            self.assertTrue(os.path.exists(path))
```
Assert that file is a file, not just that it exists.

jessebrennan replied:
Good call.

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,82 @@
+
+    @patch('hca.dss.DSSClient._download_file', side_effect=_fake_download)
+    def test_manifest_download_partial(self, _):
+        _touch_file(self.dss._file_path(self.manifest[1][4]))
```
These magic column indices are annoying but eliminating them would require a more involved refactoring.

jessebrennan replied:
I can refactor to use context manager if you want.

hannes-ucsc replied:
Not sure how a CM would help here. Maybe using a `namedtuple` for the manifest rows would help. We could even use that namedtuple in the production code to avoid these magic  constants. You could use https://docs.python.org/2/library/collections.html#collections.somenamedtuple._fields to generate the header row.

You call, honestly.

hannes-ucsc replied:
Your call if and how you want to do this.

jessebrennan replied:
Sorry I was looking at the wrong line. Yes I would like to just leave this if you don't mind.

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,82 @@
+
+    @patch('hca.dss.DSSClient.DIRECTORY_NAME_LENGTHS', [1, 2, 3])
+    def test_file_path(self):
+        self.assertRaises(AssertionError, self.dss._file_path, 'a')
```
I'd do both `abcdefghi` and `abcdefghijk`, one too few and one too many.

jessebrennan commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,82 @@
+    try:
+        os.makedirs(os.path.split(path)[0])
+    except OSError as e:
+        if 'File exists' in str(e):
```
Changed.

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,86 @@
+            '.' + os.sep + 'ad' + os.sep + '3fc1' + os.sep + 'ad3fc1e4898e0bce096be5151964a81929dbd2a92bd5ed56a39a8e133053831d',
+            '.' + os.sep + '8f' + os.sep + '3507' + os.sep + '8f35071eaeedd9d6f575a8b0f291daeac4c1dfdfa133b5c561232a00bf18c4b4',
+            '.' + os.sep + '8f' + os.sep + '3404' + os.sep + '8f3404db04bdede03e9128a4b48599d0ecde5b2e58ed9ce52ce84c3d54a3429c'
+        }
```
* Nested comprehension become harder to read as one liners.
* `os.path.join()` accepts `*args`
* `assertSetEquals` is [dicouraged](https://docs.python.org/3/library/unittest.html#unittest.TestCase.addTypeEqualityFunc)

> The list of type-specific methods automatically used by assertEqual() are summarized in the following table. Note that it’s usually not necessary to invoke these methods directly.

So:

```diff
@@ -45,14 +55,16 @@
         shutil.rmtree(self.tmp_dir)
 
     def _assert_all_files_downloaded(self):
-        files_present = {os.path.join(dir_path, f) for dir_path, _, files in scandir.walk('.') for f in files}
+        files_present = {os.path.join(dir_path, f)
+                         for dir_path, _, files in scandir.walk('.')
+                         for f in files}
         files_expected = {
-            '.' + os.sep + 'manifest.tsv',
-            '.' + os.sep + 'ad' + os.sep + '3fc1' + os.sep + 'ad3fc1e4898e0bce096be5151964a81929dbd2a92bd5ed56a39a8e133053831d',
-            '.' + os.sep + '8f' + os.sep + '3507' + os.sep + '8f35071eaeedd9d6f575a8b0f291daeac4c1dfdfa133b5c561232a00bf18c4b4',
-            '.' + os.sep + '8f' + os.sep + '3404' + os.sep + '8f3404db04bdede03e9128a4b48599d0ecde5b2e58ed9ce52ce84c3d54a3429c'
+            os.path.join('.', 'manifest.tsv'),
+            os.path.join('.', 'ad', '3fc1', 'ad3fc1e4898e0bce096be5151964a81929dbd2a92bd5ed56a39a8e133053831d'),
+            os.path.join('.', '8f', '3507', '8f35071eaeedd9d6f575a8b0f291daeac4c1dfdfa133b5c561232a00bf18c4b4'),
+            os.path.join('.', '8f', '3404', '8f3404db04bdede03e9128a4b48599d0ecde5b2e58ed9ce52ce84c3d54a3429c')
         }
-        self.assertSetEqual(files_present, files_expected)
+        self.assertEqual(files_present, files_expected)
 
     @patch('hca.dss.DSSClient.DIRECTORY_NAME_LENGTHS', [1, 2, 3])
     def test_file_path(self):
```

jessebrennan replied:
Also duh. This is much better. I forgot it took `*args`. Good to know about `assertSetEquals`.

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,86 @@
+        if e.errno == errno.EEXIST:
+            pass
+    with open(path, 'w'):
+        pass
```
```diff
@@ -16,6 +16,8 @@
     except OSError as e:
         if e.errno == errno.EEXIST:
             pass
+        else:
+            raise
     with open(path, 'w'):
         pass
 ```

jessebrennan replied:
duh! :man_facepalming:

hannes-ucsc commented on test/unit/test_manifest_download.py:
```diff
@@ -0,0 +1,86 @@
+        ['a_uuid', '1_version', 'somestuff', 'a_file_name', 'ad3fc1e4898e0bce096be5151964a81929dbd2a92bd5ed56a39a8e133053831d', '12', 'af_uuid', 'af_version', 'False'],
+        ['b_uuid', '1_version', 'somestuff', 'b_file_name', '8f35071eaeedd9d6f575a8b0f291daeac4c1dfdfa133b5c561232a00bf18c4b4', '2', 'bf_uuid', 'af_version', 'False'],
+        ['c_uuid', '1_version', 'somestuff', 'c_file_name', '8f3404db04bdede03e9128a4b48599d0ecde5b2e58ed9ce52ce84c3d54a3429c', '41', 'cf_uuid', 'af_version', 'False'],
+    ]
```
```diff
@@ -25,12 +27,20 @@
 
 
 class TestManifestDownload(unittest.TestCase):
-    manifest = [
-        ['bundle_uuid', 'bundle_version', 'file_content_type', 'file_name', 'file_sha256', 'file_size', 'file_uuid', 'file_version', 'file_indexed'],
-        ['a_uuid', '1_version', 'somestuff', 'a_file_name', 'ad3fc1e4898e0bce096be5151964a81929dbd2a92bd5ed56a39a8e133053831d', '12', 'af_uuid', 'af_version', 'False'],
-        ['b_uuid', '1_version', 'somestuff', 'b_file_name', '8f35071eaeedd9d6f575a8b0f291daeac4c1dfdfa133b5c561232a00bf18c4b4', '2', 'bf_uuid', 'af_version', 'False'],
-        ['c_uuid', '1_version', 'somestuff', 'c_file_name', '8f3404db04bdede03e9128a4b48599d0ecde5b2e58ed9ce52ce84c3d54a3429c', '41', 'cf_uuid', 'af_version', 'False'],
-    ]
+    manifest = list(zip(
+        ('bundle_uuid', 'a_uuid', 'b_uuid', 'c_uuid'),
+        ('bundle_version', '1_version', '1_version', '1_version'),
+        ('file_content_type', 'somestuff', 'somestuff', 'somestuff'),
+        ('file_name', 'a_file_name', 'b_file_name', 'c_file_name'),
+        ('file_sha256',
+         'ad3fc1e4898e0bce096be5151964a81929dbd2a92bd5ed56a39a8e133053831d',
+         '8f35071eaeedd9d6f575a8b0f291daeac4c1dfdfa133b5c561232a00bf18c4b4',
+         '8f3404db04bdede03e9128a4b48599d0ecde5b2e58ed9ce52ce84c3d54a3429c'),
+        ('file_size', '12', '2', '41'),
+        ('file_uuid', 'af_uuid', 'bf_uuid', 'cf_uuid'),
+        ('file_version', 'af_version', 'af_version', 'af_version'),
+        ('file_indexed', 'False', 'False', 'False'),
+    ))
 
     def setUp(self):
         self.prev_wd = os.getcwd()
```

hannes-ucsc replied:
I find that more readable than a 169 character line. Then again some people don't know what zip() does and might get confused. This is a suggestion.

jessebrennan replied:
Ok, I see your point.

jessebrennan replied:
Suggestion taken.
